### PR TITLE
enable write barrier verification for JIT code

### PR DESCRIPTION
**Pull request by leirocks**

### CODE REVIEW

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test OSX daily_osx_osx_debug_static please

File: lib/Backend/Lower.cpp [hunk 1]
```diff
@@ -100,6 +100,40 @@ Lowerer::Lower()
...
+        // should do this in LowererMD::GenerateWriteBarrier, however, can't insert call instruction there
+        IR::Instr* instr = m_func->m_headInstr;
+        while (instr != m_func->m_tailInstr)
+        {
+            if (instr->m_src1 && instr->m_src1->IsAddrOpnd())
+            {
+                IR::AddrOpnd* addrOpnd = instr->m_src1->AsAddrOpnd();
+                if (addrOpnd->GetAddrOpndKind() == IR::AddrOpndKindWriteBarrierCardTable)
+                {
+                    auto& lea = instr->m_prev->m_prev;
```
Comment (ThomsonTan) [inline lib/Backend/Lower.cpp]: >lea [](start = 26, length = 3)

Name it as leaInstr and shrInstr to be more consistent with other code?

File: lib/Backend/Lower.cpp [hunk 2]
```diff
@@ -19973,7 +20007,7 @@ Lowerer::GenerateFastStFld(IR::Instr * const instrStFld, IR::JnHelperMethod help
         {
             labelNext = IR::LabelInstr::New(Js::OpCode::Label, this->m_func, isHelper);
             lastBranchToNext = LowererMD::GenerateLocalInlineCacheCheck(instrStFld, typeOpnd, opndInlineCache, labelNext);
-            LowererMD::GenerateStFldFromLocalInlineCache(instrStFld, opndBase, opndSrc, opndInlineCache, labelFallThru, true);
+            this->GetLowererMD()->GenerateStFldFromLocalInlineCache(instrStFld, opndBase, opndSrc, opndInlineCache, labelFallThru, true);
```
Comment (ThomsonTan) [inline lib/Backend/Lower.cpp]: >this->GetLowererMD()-> [](start = 12, length = 22)

Why cannot use LowererMD:: to do static call?

**Comment by jianchun** [comment]
>     Lowerer::InsertMove(opndDst, opndSrc, insertBeforeInstr);

Reading this, doesn't it generate WB twice? (`InsertMove` will go to the function above `ChanteToWriteBarrierAssign` and also generate WB?)

---
Refers to: lib/Backend/LowerMDShared.cpp:4736 in 381a6b0. [](commit_id = 381a6b07cb6e4a36ae3b0e408e97936ffa3f42f8, deletion_comment = False)

File: lib/Backend/Lower.cpp [hunk 3]
```diff
@@ -100,6 +100,40 @@ Lowerer::Lower()
...
+
+#if DBG
+    // TODO: (leish)(swb) implement for arm
+#if defined(_M_IX86) || defined(_M_AMD64)
+    if (CONFIG_FLAG(ForceSoftwareWriteBarrier))
+    {
+        // find out all write barrier setting instr, call Recycler::WBSetBit for verification purpose
+        // should do this in LowererMD::GenerateWriteBarrier, however, can't insert call instruction there
+        IR::Instr* instr = m_func->m_headInstr;
+        while (instr != m_func->m_tailInstr)
```
Comment (jianchun) [inline lib/Backend/Lower.cpp]: Would `FOREACH_INSTR(instr, func->m_headInstr)` work?

**Comment by leirocks** [comment]
>     Lowerer::InsertMove(opndDst, opndSrc, insertBeforeInstr);

I saw that as well. I'm not sure why do it second time, maybe a requirement for MemRefOpnd @Hitesh can you comment on this? my change in this PR is only adding the verification bits setting

---
In reply to: [267658342](https://github.com/Microsoft/ChakraCore/pull/2222#issuecomment-267658342) [](ancestors = 267658342)

---
Refers to: lib/Backend/LowerMDShared.cpp:4736 in 381a6b0. [](commit_id = 381a6b07cb6e4a36ae3b0e408e97936ffa3f42f8, deletion_comment = False)

File: lib/Backend/Lower.cpp (hunk 2 above)
Comment (leirocks) [inline lib/Backend/Lower.cpp]: I need the lowerMD in the inner code to generate arguments loading instruction for calling my helper

File: lib/Backend/Lower.cpp (hunk 1 above)
Comment (leirocks) [inline lib/Backend/Lower.cpp]: sure, I can do it in next commit

---
In reply to: [92857326](https://github.com/Microsoft/ChakraCore/pull/2222#discussion_r92857326) [](ancestors = 92857326)

File: lib/Backend/Lower.cpp (hunk 3 above)
Comment (leirocks) [inline lib/Backend/Lower.cpp]: should work as well. do you think that's better? I can change to that pattern in next commit

---
In reply to: [92859035](https://github.com/Microsoft/ChakraCore/pull/2222#discussion_r92859035) [](ancestors = 92859035)

File: lib/Backend/Lower.cpp [hunk 4]
```diff
@@ -100,6 +100,40 @@ Lowerer::Lower()
...
+                IR::AddrOpnd* addrOpnd = instr->m_src1->AsAddrOpnd();
+                if (addrOpnd->GetAddrOpndKind() == IR::AddrOpndKindWriteBarrierCardTable)
+                {
+                    auto& lea = instr->m_prev->m_prev;
+                    auto& shr = instr->m_prev;
+                    Assert(lea->m_opcode == Js::OpCode::LEA);
+                    Assert(shr->m_opcode == Js::OpCode::SHR);
+                    m_lowererMD.LoadHelperArgument(shr, lea->m_dst);
+                    IR::Instr* instrCall = IR::Instr::New(Js::OpCode::Call, m_func);
+                    instrCall->SetSrc1(IR::HelperCallOpnd::New(IR::HelperWriteBarrierSetVerifyBit, this->m_func));
```
Comment (MikeHolman) [inline lib/Backend/Lower.cpp]: can you use m_lowererMD.ChangeToHelperCall instead?

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by dilijev** [comment]
@dotnet-bot test Ubuntu daily_ubuntu_linux_test_static please

**Comment by dilijev** [comment]
@mmitche it looks like `#include <typeinfo>` is a problem now, in addition to `<type_traits>`.
https://ci.dot.net/job/Microsoft_ChakraCore/job/master/job/daily_ubuntu_linux_test_static_prtest/75/consoleFull#-20814844619119f5c-76e5-4da8-b294-eb427d612413

**Comment by leirocks** [comment]
@dotnet-bot test OSX daily_osx_osx_debug_static please

**Pull request merged**